Editorial: Define type definitions (#870)
graphql/graphql-spec: spec/Section 3 -- Type System.md

  - InputObjectTypeDefinition

The fundamental unit of any GraphQL Schema is the type. There are six kinds
of *named type* and two kinds of *wrapped type* definitions in GraphQL.

:: A *named type* is one of the six types defined with a name: *Scalar type*,
*Enum type*, *Object type*, *Interface type*, *Union type*,
and *Input Object type*.

> ### Leaf and Composite Types

Reviewer: Good time to clarify #826 .

Author: Thanks for making this connection! That earlier phrasing is definitely confusing. I'll weigh in on that issue and make sure this PR covers the solution. 

In writing this change I found a number of cases where we confuse "value" and "type" and interchange them even in the same paragraph. We've also made the mistake of using "scalar" when we mean "leaf".

:: A *leaf type* is found at the leaves in a response tree. There are two kinds
of leaf types: *Scalar type* and *Enum type*.

The most basic *leaf type* is the *Scalar type*. A scalar represents a primitive
value, like a string or an integer. GraphQL defines a set of *built-in scalars*
and also allows a service to additionally provide any *custom scalar*.

Oftentimes, the possible values for a *leaf type* are enumerable. GraphQL offers
the *Enum type* in those cases, where the each enum type definition specifies
the set of possible values.

> :: A *composite type* compose together leaf types and other composite types as a

Author: A leaf type is scalar/enum, not including wrapped. Same with composite types being the set described here, not including wrapped.

This is good feedback on this paragraph being falsely precise - since we can compose leaf or composite types (leaf + composite = named) and any wrapped of either of those (named + wrapped(named) = all types). Thats just all types.

Going to replace with something simple:

> A *composite type* composes other types together as a set of named fields

> :: A *composite type* compose together leaf types and other composite types as a

Reviewer 2: Sounds good to me 👍

actual concrete value which will a field will resolve during execution, composed
of other *output type* fields, allowing responses of nested data.

Oftentimes it is useful to provide a *composite type* as an *input type* to a
field argument or variable; the *Input Object type* composes a set of
*input type* fields, allowing inputs of nested data.

### Abstract Types

:: An *abstract type* defines a set of possible types, one of which will be
resolved during field execution. There are two kinds of abstract types:
*Interface type* and *Union type*.

:: The type an abstract type resolves to during execution is referred to as the
*concrete Object type*, which is always an *Object type*.

An *Interface type* defines a list of fields. An type which implements this
Interface must also implement those fields. Whenever a field claims it will
return an *Interface type*, it will return an implementing
*concrete Object type* during field execution.

A *Union type* defines a list of possible types. Similar to Interfaces,
whenever a field claims it will return a *Union type*, it will return one of the
possible types as the *concrete Object type* during field execution.

> ### Wrapped Types

Author: @benjie - would love your thoughts on this newly written subsection.

This is based on [your comment](https://github.com/graphql/graphql-spec/pull/845#issuecomment-826341088) with very minor changes:

* **Wrapped type** - the kind of type including Lists and Non-Nulls
  * **List type** - changed from your suggestion of **List wrapped type**
  * **Non-Null type** - changed from your suggestion of **Non-Null wrapped type**
* **Inner type** - the type immediately wrapped (eg the `ofType`)
  * **Item type** - I added this - the inner type of a List
  * **Nullable type** - I added this - the inner type of a Non-Null

scalar string returns either {null} or a singular string.

:: A *wrapped type* wraps another type (its *inner type*) to create a collection
or variation of that type. There are two kinds of wrapped types: *List type* and
*Non-Null type*.

:: The *inner type* is the type immediately wrapped by a *wrapped type*, which
itself may be a *wrapped type* allowing for combinations of *wrapped type*. If
continuously unwrapped, ultimately a *named type* is found.

A GraphQL schema may describe that a field or argument expects a list of another
type; the *List type* is provided for this purpose, and wraps an *inner type*
describing the items in the list.

Similarly, a field or argument may describe that it must never be {null}. The
*Non-Null type* is provided for this purpose, and wraps an *inner type*
describing the expected type, if not {null}.

### Input and Output Types

Types are used throughout GraphQL to describe both the values accepted as input
to arguments and variables as well as the values output by fields. These two
uses categorize types as an *input type* or an *output type*. Some kinds of
types, like Scalar and Enum types, can be used as both input types and output

> types; other kinds of types can only be used in one or the other.

Reviewer 2: *A little off-topic* Whilst we're on this subject... generally `scalar` types are suitable for input and output, but in some cases (for example the `Upload` scalar used by the GraphQL Upload Spec) a custom scalar may only be suitable for one but not the other. Is this specifically disallowed, or do we need to be more cautious with our wording?

Author: We actually don't have the tools to assert this restriction right now. We don't specifically disallow it, but we also can't enforce it (with validation - a custom scalar could just fail coercion, but thats a bit of a hack IMO)

Reviewer 2: Yep; I’m happy keeping the status quo; just felt it was worth mentioning.

Reviewer 3: > We actually don't have the tools to assert this restriction right now

I'd be interested in digging into this distinction for a later release: I have some usages around input vs output enums that I'd like to add clarity to. As the spec stands, [this statement](https://spec.graphql.org/draft/#sel-GAHfRDABABvEomB):
> GraphQL services must return one of the defined set of possible values. If a reasonable coercion is not possible they must raise a field error.

can mean it's actually a breaking change to add a new enum value to an existing enum  type (we make no statement about avoiding exhaustive switches on output enums, for example). We don't get into details in the spec itself on what a breaking type system change is, with just a [short blurb about type system evolution](https://spec.graphql.org/draft/#sec-Validation.Type-system-evolution).

Reviewer 4: https://github.com/graphql/graphql-js/issues/968#issuecomment-317834310

Reviewer 4: Relevant

  - InputObjectTypeDefinition

The fundamental unit of any GraphQL Schema is the type. There are six kinds
of *named type* and two kinds of *wrapped type* definitions in GraphQL.

:: A *named type* is one of the six types defined with a name: *Scalar type*,
*Enum type*, *Object type*, *Interface type*, *Union type*,
and *Input Object type*.

### Leaf and Composite Types

The *named type* definitions are divided into leaf types and composite types,
which together can compose to represent complex type hierarchies.

:: A *leaf type* is found at the leaves in a response tree. There are two kinds
of leaf types: *Scalar type* and *Enum type*.

The most common *leaf type* is the *Scalar type*. GraphQL defines a set of
*built-in scalars* for representing common primitive values and also allows a

> service to additionally provide a *custom scalar type* to represent any value.

Reviewer 2: I'm closing this, we're consistently "custom scalar type" everywhere.

ScalarTypeDefinition : Description? scalar Name Directives[Const]?

> :: A *Scalar type* represents primitive leaf values in a GraphQL type system.
> GraphQL responses take the form of a hierarchical tree; the leaves of this tree
> are typically GraphQL Scalar types (but may also be Enum types or {null} values).

Reviewer 2: When we're talking about `types` I'm not sure it makes sense to include `{null}`? This might imply you could have `type A {foo: null}`. We should also factor in that the leaves may be lists/non-nulls:

```suggestion
:: A *Scalar type* represents primitive leaf values in a GraphQL type system.
GraphQL responses take the form of a hierarchical tree; the *named type* of the
leaves of this tree are typically GraphQL Scalar types (but may also be Enum
types).
```

Author: Agreed. This was a case where we were mixing "types" and "values". I'll rewrite this

IsValidImplementationFieldType(fieldType, implementedFieldType):
  1. If {fieldType} is a *Non-Null type*:
     1. Let {nullableType} be the *nullable type* of {fieldType}.
     2. Let {implementedNullableType} be the *nullable type*
        of {implementedFieldType} if it is a *Non-Null type*, otherwise let it
        be {implementedFieldType} directly.
     3. Return {IsValidImplementationFieldType(nullableType, implementedNullableType)}.
  2. If {fieldType} is and {implementedFieldType} are both a *List type*:
     1. Let {itemType} be the *item type* of {fieldType}.
     2. Let {implementedItemType} be the *item type* of {implementedFieldType}.
     3. Return {IsValidImplementationFieldType(itemType, implementedItemType)}.
  3. If {fieldType} is the same type as {implementedFieldType} then return {true}.
  4. If {fieldType} is an *Object type* and {implementedFieldType} is
     a *Union type* and {fieldType} is a possible type of {implementedFieldType}
     then return {true}.
  5. If {fieldType} is an *Object type* or *Interface type* and
     {implementedFieldType} is an *Interface type* and {fieldType} declares it
     implements {implementedFieldType} then return {true}.

>   6. Otherwise return {false}.

Reviewer 2: Much clearer :+1:

EnumValueDefinition : Description? EnumValue Directives[Const]?

> :: An *Enum type*, like a *Scalar type*, also represent leaf values in a GraphQL

Reviewer 2: ```suggestion
:: An *Enum type*, like a *Scalar type*, also represents leaf values in a GraphQL
```

Fields may accept arguments to configure their behavior. These inputs are often
scalars or enums, but they sometimes need to represent more complex values.

> :: An *Input object type* defines a set of input fields; the input fields are

Reviewer 2: ```suggestion
:: An *Input Object type* defines a set of input fields; the input fields are
```

**Type Validation**

1. A *Non-Null type* must not wrap another *Non-Null type*.

### Combining List and Non-Null

> The *List type* and *Non-Null types* can compose, representing more complex

Reviewer 2: You've been careful to always use the singular elsewhere; I'm not sure if this relates to how the glossary feature in spec-md works. If this need not be singular on these technical grounds then plural is fine :+1: 

```suggestion
The *List type* and *Non-Null type* can compose, representing more complex
```
